### ISSUE TITLE
Wrong device when using device="cpu" with torch.device

### ISSUE THREAD

**Issue by chrisdesa** [huggingface/safetensors#503]
### System Info

Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

- `transformers` version: 4.43.1
- Platform: Linux-5.15.0-112-generic-x86_64-with-glibc2.35
- Python version: 3.12.2
- Huggingface_hub version: 0.23.2
- Safetensors version: 0.4.3
- Accelerate version: 0.33.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.4.0.dev20240513+cu121 (True)
- GPU type: NVIDIA H100 80GB HBM3

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Reproduction

```
with torch.device('cuda:0'):
    safetensors.torch.load_file('filename.safetensors',device='cpu')['key'].device
```

### Expected behavior

The expected behavior is for the `device='cpu'` argument to override the torch default and load the tensors on the CPU, but the actual behavior is it loads the tensor onto the `'cuda:0'` GPU. The `device='cpu'` argument seems to be interpreted as "whatever the default device is right now" and not as the actual CPU.

**Comment by Narsil** [comment]
I understand the concern, however I do not think this should be fixed.

`with torch.device(x)` 
Changes the default allocation place, which messes a lot of internal logic in `safetensors` (it doesn't make it *wrong* it makes it highly inefficient, since your first allocating on device before moving back to the CPU which about the slowest implementation possible).

The behavior *is* changeable, by being extremely defensive about those global modifiers that are the context managers. However as a general rule, I'd avoid using any rules depending on torch internals, since they tend to change a lot (this context manager didn't exist when this library was created).

Also, in the future, there might be ways to skip CPU allocation entirely meaning the internals would have to be rewritten again, introducing potential breaking changes (which is highly avoided in this library).

Given all these considerations, I'm marking this as wontfix, and instead encourage you and other users to not use contradicting device locations from a context manager.

So:

Either
```python
with torch.device("cuda:0"):
    # Fix the context manager, that make everything much faster.
    with torch.device("cpu"):
      weights = load_file(filename)
```

Or 
```python
# Remove the context manager.
weights = load_file(filename, device="cpu")
```

Just to note that 
```python
with torch.device("cuda:0"):
   weights = load_file(filename)
```
is totally valid intent behavior and will work transparently currently (contrary to what would happen should be defensive).

**Comment by sdake** [comment]
I agree with you broken designs (context managers have all kinds of problems, and we are inventing high performance compute here, where cycles matter), so its better to say no.

How viable would a deprecation warning be? If context manager -> poof.

**Comment by Narsil** [comment]
For `torch` we are using torch's internals to load tensors, which afaik is hard to beat without completely sidestepping the torch engine (which will cause issues with the torch allocator most likely).

`torch.to(device, non_blocking=True)` works with safetensors.
CPU load actually doesn't allocate since it's memory mapped. Until it's read, but you can copy before that.
I've tiptoed with that in previous iterations of torch, but their primitive/allocators had significant bugs when using async alloc/memcpy.

Maybe they have fixed it since.

Thanks for the kind words, and I also highly respect your work in vLLM.
